compiler/il/OMRBlock.cpp

```diff
@@ -1207,6 +1213,23 @@ static TR::SymbolReference * createSymRefForNode(TR::Compilation *comp, TR::Reso
  * Splits the blocks from the split point manually handling the GlRegDeps created by the Global Register Allocator.
  * If there are available global registers which can be used, it will utililze them to uncommon nodes across split points.
  * 
+ * @details
+ * It uses following simplified algorithm to split blocks and uncommon nodes. 
+ * 1. Analyze trees before the split point to gather information about nodes which are referenced in new block and needs replacement after split.
+ * 2. Now as it has information about all the nodes which are required to be uncommoned, it starts cretaing replacment nodes as per following rules.
+ *    a. If node to be uncommoned is regLoad / load of constant then replacement node will be copy of the original node and also mark the used register unavailable. 
+ *    b. If node to be uncommoned is stored into a register after split point and that register is available, then replacement node will be regLoad from that
+ *       register. It also makes sure that regStore tree is moved to before split point also marks used register unavailable.
+ *    c. If node to be uncommoned is stored into a register before split point and node does not succeeds in [b] case and register is available then replacement
+ *       node will be regLoad from that register also marks used register unavailable.
+ * 3. While checking for case-b and case-c from Step 2 it also checks regStores and PassThrough nodes after split point. 
+ *    a. If node under pass through requires to be uncommoned and uses different register then the replacement node
```

Comment:
I don't understand what is meant by "of the node under pass through requires to be uncommoned" here.

Reply:
Basically this comment is for following kind of case. 

```
n1n BBStart
n2n    GlRegDeps
n3n       aRegLoad R5
...
n4n    iRegStore R5
n5n       nodeY
...
// Split Point
...
n6n    <Branch Node>
          ...
n7n       GlRegDeps ()
n8n          PassThrough R5
n5n             nodeY
...
n6n nodeX
n3n    => aRegLoad
```
In this case we have an exit edge which uses `R5` for nodeY which is already in use by `n3n`.  In this case, as `n4n`(Corresponding regStore to `n8n` ) is seperated by split point, it needs to add `iRegStore` in newly created block before the `n6n`.

Reply:
Can you please show what the transformed code would look like in this case ?

Reply:
To highlight the scenario, extracting trees from one of the log file I have. 
```
n1n BBStart
n2n     GlRegDeps
n3n         aRegLoad R1 <temp slot 1>
n4n ResolveCHK
n5n     aloadi <some field>
n6n         aload <temp slot 2>
...
n7n aRegStore R2
n5n     => aloadi
n8n aRegStore R3
n5n     => aloadi
...
// SplitPoint
n9n ifcmpeq goto block_x
    ...
n10n    GlRegDeps ()
n3n         => aRegLoad
n11n        PassThrough R3
n5n             => aloadi
n12n BBEnd
...
n13n BBStart (Extention of previous block)
...
n14n ifcmpeq goto block_y
     ...
n15n    GlRegDeps ()
n3n         => aRegLoad
n16n        PassThrough R2
n5n             => aloadi
...
```
In above case GRA has chosen two registers for node `n5n` for multiple exit edges in given
extended basic block and stores the node into  register before split point.  
For this case, it will transform above trees as following. 
```
n1n BBStart
n2n     GlRegDeps
n3n         aRegLoad R1 <temp slot 1>
n4n ResolveCHK
n5n     aloadi <some field>
n6n         aload <temp slot 2>
...
n7n aRegStore R2 <slotX>
n5n     => aloadi
n8n aRegStore R3 <slotY>
n5n     => aloadi
...
n17n BBStart (New Created Block)
n18n    GlRegDeps
n19n        aRegLoad R1 <temp slot 1>
n20n        aRegLoad R3 <slotY>
// Trees after split point
n9n ifcmpeq goto block_x
    ...
n10n    GlRegDeps ()
n19n         => aRegLoad
n11n        PassThrough R3
n20n             => aRegLoad
n12n BBEnd
...
n13n BBStart (Extention of previous block)
...
n21n aRegStore R2 <sloyX>
n20n    => aRegLoad
n14n ifcmpeq goto block_y
     ...
n15n    GlRegDeps ()
n3n         => aRegLoad
n16n        PassThrough R2
n20n             => aRegLoad
...
```
So whichever exit edge expects `n5n` in R2 will have similar trees created as `n21n`. 
I have changed my example from https://github.com/eclipse/omr/pull/4152#discussion_r311647176 as I am not sure if the example I gave is functionally correct. But even if that case is possible, it will give `n3n aRegLoad R5` precedence and create following trees near the `GlRegDeps` 
```
...
n9n aRegStore R5
n10n   replacement of nodeY
n6n <Branch Node>
           ...
n7n        GlRegDeps()
n8n            PassThrough R5
n10n              => nodeY
...
```

Reply:
Given n7n, namely aRegStore into R2, why don't we have an aRegLoad of R2 under the GlRegDeps for the newly created block's BBStart ? If we'd have had that, then the aRegLoad of R2 would have been the child of the GlRegDeps n15n instead of n16n.

Reply:
I guess for this I would have to use `multi-map` to represent multiple replacement nodes for single key node and then while analyzing the `GlRegDeps` at the exit edges after split point, I could have done what you are suggesting. 
Also I do not have the log file handy, but while running Spark/HiBench on Linux On Z, I encounter a case which I mentioned in https://github.com/eclipse/omr/pull/4152#discussion_r311647176 in which case it will create duplicate register dependency where at the start of the new block. we will have two different regLoads of two different symbols, using the same register. 
Because of these reasons, we chose simplest option here to make sure we get this functionally working first. Then we can enhance this to minimize register shuffling.